### PR TITLE
Fluent api

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -219,9 +227,7 @@ By default, Limestone API fetches data from the Limestone cache layer. It works
 We strongly recommend using the default fetching mechanism. But if you want to fetch data directly from Arweave you can do it by initialising a new `LimestoneApi` client and setting `useCache` option to `false`.
```
Comment (jakub-wojciechowski) [inline README.md]: We strongly recommend using the default fetching mechanism which leverages cache to speed up queries.

File: README.md [hunk 2]
```diff
@@ -232,11 +238,8 @@ console.log(price.value); // AR price value fetched directly from Arweave
 
 ----------------------------------------------
 
-## 📅 Roadmap
-
 ### Fluent Interface
-We currently work on a fluent interface to make Limestone API even easier to use.
-You can [see examples](examples/fluent-api/example.js) of fluent interface usage and let us know [what you think](https://discord.gg/PVxBZKFr46) (we would be very grateful 😉).
+Limestone also implements an additional fluent interface to make it even simpler to execute some price fetching queries for popular use cases. [Learn more](docs/FLUENT_INTERFACE.md)
```
Comment (jakub-wojciechowski) [inline README.md]: Limestone implements a fluent interface to simplify query creation thanks to a human readable syntax.

File: __tests__/fluent-interface.test.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    const price = await limestone.query()
+      .symbol("AR")
+      .hoursAgo(24)
+      .exec();
+
+    const timeDiff = Date.now() - 24 * 3600 * 1000 - price.timestamp;
+
+    expect(price.symbol).toBe("AR");
+    expect(price.value).toBeGreaterThan(1);
+    expect(timeDiff).toBeLessThan(MAX_TIME_DIFF);
```
Comment (jakub-wojciechowski) [inline __tests__/fluent-interface.test.ts:58]: I'll additionally check if the prices dates are within time range

File: __tests__/fluent-interface.test.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  });
+
+  test("Should get historical AR price for last 7 days", async () => {
+    const prices = await limestone.query()
+      .symbol("AR")
+      .forLastDays(7)
+      .exec();
+
+    expect(prices.length).toBeGreaterThan(165);
+    expect(prices.length).toBeLessThan(170);
```
Comment (jakub-wojciechowski) [inline __tests__/fluent-interface.test.ts]: I'll additionally check if the prices dates are within time range

File: __tests__/fluent-interface.test.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  });
+
+  test("Should get historical AR price for the last 1 day", async () => {
+    const prices = await limestone.query()
+      .symbol("AR")
+      .forLastDays(1)
+      .exec();
+
+    expect(prices.length).toBeGreaterThan(23);
+    expect(prices.length).toBeLessThan(25);
```
Comment (jakub-wojciechowski) [inline __tests__/fluent-interface.test.ts:82]: I'll additionally check if the prices dates are within time range

File: __tests__/fluent-interface.test.ts [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  });
+
+  test("Should get AR price in time range", async () => {
+    const prices = await limestone.query()
+      .symbol("AR")
+      .fromDate("2021-04-19")
+      .toDate("2021-04-20")
+      .exec();
+
+    expect(prices.length).toBe(24);
```
Comment (jakub-wojciechowski) [inline __tests__/fluent-interface.test.ts:96]: I'll additionally check if the prices dates are within time range

File: __tests__/fluent-interface.test.ts [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      .latest()
+      .exec();
+
+    expect(prices["AR"]).toBeDefined();
+    expect(prices["ETH"]).toBeDefined();
+    expect(prices["BTC"]).toBeDefined();
+    expect(prices["AR"].value).toBeGreaterThan(0.1);
+    expect(prices["ETH"].value).toBeGreaterThan(100);
+    expect(prices["BTC"].value).toBeGreaterThan(1000);
+    expect(Date.now() - prices["AR"].timestamp).toBeLessThan(MAX_TIME_DIFF);
```
Comment (jakub-wojciechowski) [inline __tests__/fluent-interface.test.ts:119]: You've forgotten to check ETH timestamp ;)

File: __tests__/get-historical-price.test.ts [hunk 8]
```diff
@@ -8,45 +8,45 @@ const shouldNotHaveTechProps = (price: any) => {
...
     const symbol = "AR";
-    const date = new Date("2021-03-14");
+    const date = new Date("2021-04-17");
     const price: any =
       await limestone.getHistoricalPrice(symbol, { date });
 
     expect(price).toBeDefined();
     expect(price.symbol).toBe(symbol);
-    expect(price.value).toBe(13.98);
+    expect(price.value).toBeCloseTo(25.923827794046517);
```
Comment (jakub-wojciechowski) [inline __tests__/get-historical-price.test.ts]: Hmm, shouldn't we increase the precision of toBeCloseTo(x,precision)

File: __tests__/get-historical-price.test.ts [hunk 9]
```diff
@@ -8,45 +8,45 @@ const shouldNotHaveTechProps = (price: any) => {
...
-  test("Should get AR price for 2021-03-14", async () => {
+  test("Should get AR price for 2021-04-17", async () => {
     const symbol = "AR";
-    const date = new Date("2021-03-14");
+    const date = new Date("2021-04-17");
     const price: any =
       await limestone.getHistoricalPrice(symbol, { date });
 
     expect(price).toBeDefined();
     expect(price.symbol).toBe(symbol);
```
Comment (jakub-wojciechowski) [inline __tests__/get-historical-price.test.ts:18]: Let's also verify the price timestampt

File: docs/FLUENT_INTERFACE.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
+# Fluent interface for limestone
+
+Limestone API supports a fluent interface to make the price fetching even simpler.
```
Comment (jakub-wojciechowski) [inline docs/FLUENT_INTERFACE.md]: Limestone implements a fluent interface to simplify query creation thanks to a human readable syntax.

File: src/limestone-query.ts [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        ) {
+          const diff = getTimeDiff(startDate, endDate);
+          if (diff >= 24 * 3600 * 1000) {
+            this.params.interval = 3600 * 1000;
+          } else {
+            this.params.interval = 1;
+          }
+        }
+
+        // TODO: check types
```
Comment (jakub-wojciechowski) [inline src/limestone-query.ts]: Let's put a task in the backlog to improve types handling in fluent interface.

File: src/signature-verifier.ts [hunk 12]
```diff
@@ -22,8 +22,9 @@ export default class SignatureVerifier {
       signerPublicKey: publicKey,
     });
 
-    const addressFromPublicKey =
-      await this.arweaveProxy.arweaveClient.wallets.ownerToAddress(publicKey);
+    const addressFromPublicKey = await this.arweaveProxy.arweaveClient.wallets.ownerToAddress(
+      publicKey,
```
Comment (jakub-wojciechowski) [inline src/signature-verifier.ts:26]: Hmm, what's the purpose of "," ?

**Review by jakub-wojciechowski — CHANGES_REQUESTED** [review]
Looks really good!

File: src/signature-verifier.ts (hunk 12 above)
Comment (hatskier) [reply src/signature-verifier.ts:26]: It was added by our code formatter

File: src/limestone-query.ts (hunk 11 above)
Comment (hatskier) [reply src/limestone-query.ts]: Added

**Pull request merged**